src/devices/Card/Ndef/samples/Program.cs

```diff
@@ -0,0 +1,443 @@
+        var ret = mifareCard.RunMifareCardCommand();
+        if (ret < 0)
+        {
+            // This will reselect the card in case of issue
+            mifareCard.ReselectCard();
+            // Try another one
+            mifareCard.Command = MifareCardCommand.AuthenticationB;
+            ret = mifareCard.RunMifareCardCommand();
+        }
+
+        if (ret >= 0)
+        {
+            mifareCard.BlockNumber = block;
+            mifareCard.Command = MifareCardCommand.Read16Bytes;
+            ret = mifareCard.RunMifareCardCommand();
+            if ((ret >= 0) && (mifareCard.Data is object))
+            {
+                Console.WriteLine($"Bloc: {block}, Data: {BitConverter.ToString(mifareCard.Data)}");
+            }
+            else
+            {
+                Console.WriteLine($"Error reading bloc: {block}");
+            }
+
+            if ((block % 4 == 3) && (mifareCard.Data is object))
+            {
+                // Check what are the permissions
+                for (byte j = 3; j > 0; j--)
+                {
+                    var access = mifareCard.BlockAccess((byte)(block - j), mifareCard.Data);
+                    Console.WriteLine($"Bloc: {block - j}, Access: {access}");
+                }
+
+                var sector = mifareCard.SectorTailerAccess(block, mifareCard.Data);
+                Console.WriteLine($"Bloc: {block}, Access: {sector}");
+            }
+        }
+        else
+        {
+            Console.WriteLine($"Authentication error");
+        }
+    }
+}
+
+void ReadNdef(MifareCard mifareCard)
+{
+    Console.WriteLine("Reading Mifare card content");
+    mifareCard.TryReadNdefMessage(out NdefMessage message);
+
+    if (message.Records.Count == 0)
+    {
+        Console.WriteLine("Sorry, there is no NDEF message in this card or I can't find them");
+    }
+
+    foreach (var msg in message.Records)
+    {
+        Console.WriteLine("Record header:");
+        Console.WriteLine($"  Is first message: {msg.Header.IsFirstMessage}, is last message: {msg.Header.IsLastMessage}");
+        Console.Write($"  Type name format: {msg.Header.TypeNameFormat}");
+        if (msg.Header.PayloadType is object)
+        {
+            Console.WriteLine($", Payload type: {BitConverter.ToString(msg.Header.PayloadType)}");
+        }
+        else
+        {
+            Console.WriteLine();
+        }
+
+        Console.WriteLine($"  Is composed: {msg.Header.IsComposedMessage}, is Id present: {msg.Header.MessageFlag.HasFlag(MessageFlag.IdLength)}, Id Length value: {msg.Header.IdLength}");
+        Console.WriteLine($"  Payload Length: {msg.Payload?.Length}, is short message= {msg.Header.MessageFlag.HasFlag(MessageFlag.ShortRecord)}");
+
+        if (msg.Payload is object)
+        {
+            Console.WriteLine($"Payload: {BitConverter.ToString(msg.Payload)}");
+        }
+        else
+        {
+            Console.WriteLine("No payload");
+        }
+
+        if (UriRecord.IsUriRecord(msg))
```

Comment:
consider message having enum "RecordType" and then switching here based on that enum rather than IsXyz, IsXyv, IsFoo, ...

Reply:
Yes, good idea, I'll add this.

Reply:
Well, that won't help, you still need to check what kind of message it is. The enum won't help in fact.

Reply:
I am checking is it's a valid Xyz inside Xyz because the base NdefRecord has no idea of the kind of message it is